nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch.java

```diff
@@ -0,0 +1,329 @@
+            .required(true)
+            .expressionLanguageSupported(false)
+            .build();
+    public static final PropertyDescriptor SPLIT_UP_AGGREGATIONS = new PropertyDescriptor.Builder()
+            .name("el-rest-split-up-aggregations")
+            .displayName("Split up aggregation results")
+            .description("Split up aggregation results into one flowfile per result.")
+            .allowableValues(SPLIT_UP_HITS_NO, SPLIT_UP_YES)
+            .defaultValue(SPLIT_UP_HITS_NO.getValue())
+            .required(true)
+            .expressionLanguageSupported(false)
+            .build();
+
+    public static final PropertyDescriptor CLIENT_SERVICE = new PropertyDescriptor.Builder()
+            .name("el-rest-client-service")
+            .displayName("Client Service")
+            .description("An ElasticSearch client service to use for running queries.")
+            .identifiesControllerService(ElasticSearchClientService.class)
+            .required(true)
+            .build();
+
+    private static final Set<Relationship> relationships;
+    private static final List<PropertyDescriptor> propertyDescriptors;
+
+    private volatile ElasticSearchClientService clientService;
+
+    static {
+        final Set<Relationship> _rels = new HashSet<>();
+        _rels.add(REL_ORIGINAL);
+        _rels.add(REL_FAILURE);
+        _rels.add(REL_HITS);
+        _rels.add(REL_AGGREGATIONS);
+        relationships = Collections.unmodifiableSet(_rels);
+
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        descriptors.add(QUERY);
+        descriptors.add(QUERY_ATTRIBUTE);
+        descriptors.add(INDEX);
+        descriptors.add(TYPE);
+        descriptors.add(CLIENT_SERVICE);
+        descriptors.add(SPLIT_UP_HITS);
+        descriptors.add(SPLIT_UP_AGGREGATIONS);
+
+        propertyDescriptors = Collections.unmodifiableList(descriptors);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        clientService = context.getProperty(CLIENT_SERVICE).asControllerService(ElasticSearchClientService.class);
+    }
+
+    @OnStopped
+    public void onStopped() {
+        this.clientService = null;
+    }
+
+
+    private final ObjectMapper mapper = new ObjectMapper();
+
+    private String getQuery(FlowFile input, ProcessContext context, ProcessSession session) throws IOException {
+        String retVal = null;
+        if (context.getProperty(QUERY).isSet()) {
+            retVal = context.getProperty(QUERY).evaluateAttributeExpressions(input).getValue();
+        } else if (input != null) {
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            session.exportTo(input, out);
```

Comment:
Looks like if the Query property isn't set and there is an incoming FF then it will use the FF's content as the query. I don't see this documented in the Query props. This should at least be documented and also ideally, there would be some custom validation too.

Reply:
Yeah, it needs the documentation. That was a recent addition. I'm not sure about the validation because it's expecting a JSON payload, and if that's malformed in any respect the client service will throw an exception back to the processor now.